[JOJI-702] 페이지네이션

MuteAim opened the pull request:
- 현재까지 구현된 페이지네이션입니다.
- 페이지 그룹은 5페이지씩 존재합니다.
- <,>버튼은 5페이지 단위로 페이지그룹을 이동합니다
![image](https://github.com/user-attachments/assets/188cc3fd-78f4-47e5-9d63-63e2129ec12f)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **새로운 기능**
	- 공지사항 페이지에 페이지네이션 기능 추가
	- 페이지 간 이동 및 현재 페이지 표시 가능
	- 페이지네이션 컴포넌트를 통한 직관적인 내비게이션 제공

- **버그 수정**
	- 컴포넌트 및 훅의 디렉토리 구조 개선
	- 가독성과 유지보수성 향상을 위한 코드 리팩토링

- **기타 변경사항**
	- 프로젝트 디렉토리 구조 전면 재구성
	- 컴포넌트 모듈화 및 재사용성 증대
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MuteAim committed:
feat: 페이지네이션 구현

MuteAim committed:
feat: 페이지네이션 개선

MuteAim committed:
feat: 페이지네이션 개선

-페이지네이션 페이지그룹 이동 개선

AlangGY commented on src/app/post/page.js:
```diff
@@ -1,30 +1,23 @@
   useEffect(() => {
     const fetchPosts = async () => {
+      setLoading(true);
       const response = await fetch(`${process.env.NEXT_PUBLIC_BASE_URL}/posts`);
```
현재 `/posts` api는 페이지네이션을 지원하고 있어요.

https://easily-api.jojicompany.com/api#/%EA%B2%8C%EC%8B%9C%EA%B8%80/PostsController_getAllPosts

를 참고해서, 원하는 게시글의 수와 페이지를 로드해올 수 있도록 처리해주세요

AlangGY commented on src/app/post/page.js:
```diff
@@ -92,7 +117,38 @@ export default function PostPage() {
+        >
+          &gt;
+        </Button>
+      </div>
```
shadcn에 Pagination 이라는 컴포넌트가 있습니다.
우선 쓱 보기에는 그쪽을 쓰는편이 이쁠것 같으니 그것으로 적용해주세요 ㅎ

https://ui.shadcn.com/docs/components/pagination#

AlangGY commented on src/app/post/page.js:
```diff
@@ -37,6 +30,44 @@ export default function PostPage() {
   }, []);
 
+  const totalPages = Math.ceil(posts.length / postSize);
+
+  const currentPosts = posts.slice(
+    (currentPage - 1) * postSize,
+    currentPage * postSize
+  );
+
+  const handlePageChange = (page) => {
+    if (page >= 1 && page <= totalPages) {
+      setCurrentPage(page);
+    }
+  };
+
+  // 이전 페이지 그룹으로 이동
+  const handlePrev = () => {
+    const prevGroupStart = Math.max(Math.floor((currentPage - 1) / 5) * 5, 1);
+    setCurrentPage(prevGroupStart); 
+  };
+  
+  // 다음 그룹으로 이동
+  const handleNext = () => {
+    const nextGroupStart = Math.min(
+      Math.floor((currentPage - 1) / 5) * 5 + 6, totalPages);
+    setCurrentPage(nextGroupStart); 
+  };
+
+  const getPageRange = () => {
+    const pages = [];
+    const rangeStart = Math.floor((currentPage - 1) / 5) * 5 + 1; // 5단위로 시작 페이지 계산
+    const rangeEnd = Math.min(rangeStart + 4, totalPages); // 끝 페이지는 totalPages를 넘지 않도록 설정
+
+    for (let i = rangeStart; i <= rangeEnd; i++) {
+      pages.push(i);
+    }
+
+    return pages;
+  };
+
```
페이지 처리와 관련된 로직을 별도 custom hook으로 빼주면 좋을것 같네용.

제가 대시보드 개발시 만들어두었던 [usePagination](https://github.com/joji-company-dev/easily/blob/main/modules/common/hooks/usePagination.tsx)을 참고해서, 정리를 한번 해주시면 좋겠습니다 ㅎ

AlangGY replied:
usePagination의 [사용처 코드](https://github.com/joji-company-dev/easily/blob/main/modules/community/components/PostsTable.tsx) 예시

AlangGY reviewed (changes requested):
작업하시느라 고생많으셨습니다!

수정이 되어야할것 같다 싶은 부분을 코멘트로 남겨두었으니 확인해보세요~

dohee12 reviewed (approved):
멋집니다 ~~ 나중에 공부할 때도 참고해보기 좋을 것 같아요

MuteAim committed:
feat: shadcn pagination 설치

MuteAim committed:
Merge branch 'development' into feat/JOJI-702/페이지네이션

MuteAim committed:
feat: shadcn을 적용한 페이지네이션 구현

-한 페이지에 보여줄 NOTICE 게시글을 정하고 싶으면 limit= 에다가 원하는 숫자를 입력하면됩니다.

MuteAim committed:
feat: 페이지이동 버튼 이름변경 + 기능추가

-1페이지 일때는 이전페이지를 잠구고 마지막 페이지일때는 다음버튼을 잠구도록 설정했습니다.

MuteAim committed:
feat: usePaginaion 커스텀 훅 구현

- currentPage: 현재페이지를 반환
- totalPages: 총 페이지 수를 반환
- setTotalPages: totalPages 값을 변경하는 함수. 외부에서 총 페이지 수를 설정할 수 있음
- handlePageChange: 페이지를 변경하는 함수
- goToNextPage: 다음 페이지로 이동하는 함수
- goToPreviousPage: 이전 페이지로 이동하는 함수
- getPageRange: 페이지 범위를 계산하여 반환하는 함수

AlangGY commented on src/app/_components/hooks/usePaigination.js:
파일명에 오타가 있습니다! `usePaigination` -> `usePagination` 으로 파일명 변경해주세용

MuteAim replied:
수정했습니돱

AlangGY reviewed (approved):
수정해야할꺼 하나 더 남겨두었습니다 ㅎ 그것만 반영하고 머지해주셔도 좋을것 같아요

MuteAim committed:
[JOJI-700] 디렉토리구조 재구성 및 문서화 (#19)

* docs: 디렉토리구조 재구성 및 문서화

-디렉토리 구조를 최적화하였습니다

* refactor: usePagination 오타수정

* refactor: 오타수정

MuteAim committed:
feat: 페이지 그룹이동 구현

= 사용시 주석내용처럼 호출

MuteAim committed:
Merge branch 'development' into feat/JOJI-702/페이지네이션

MuteAim committed:
refactor: 디렉토리구조 변경 및 파일명변경

MuteAim committed:
feat: usePagination 코드 수정

- 그룹이동 함수 추가
